Introduce and move new Arduino examples and configuration updates.

gojimmypi opened the pull request:
# Description
This PR introduces a new AES CTR (Counter Mode) Arduino example sketch along with supporting template files and configuration updates. The example demonstrates AES CTR encryption and decryption capabilities of wolfSSL on Arduino platforms.

edit: _Moving_ Arduino examples. See https://github.com/wolfSSL/wolfssl-examples/pull/499

edit (2): Expanded the scope of handling the various combinations of these macros in `user_settings.h`:

- `#define NO_TLS`
- `#define WOLFCRYPT_ONLY`
- `#define NO_WOLFSSL_CLIENT`
- `#define NO_WOLFSSL_SERVER`

Some combinations would cause compiler warnings about things declared and note used. Additional gating applied to handle this.

One combination now causes a hard error via `settings.h`: 

```
//#define NO_TLS
//#define WOLFCRYPT_ONLY
#define NO_WOLFSSL_CLIENT
#define NO_WOLFSSL_SERVER
```

## Key Changes

1. **New AES CTR Example**:
   - Added IDE/ARDUINO/sketches/wolfssl_AES_CTR/wolfssl_AES_CTR.ino - A complete example demonstrating AES CTR encryption/decryption
   - Added supporting project files (README.md, .adeps, .sln, .vgdbproj) for various Arduino IDE environments
   - Tested on Arduino using Teensy 4.1 (ARM Cortex-M7 at 600 MHz)
   - AES CTR GPLV2 Contributed by Chris Jennings Feb 2025; Thank you!

2. **Arduino Template Example**:
   - Added IDE/ARDUINO/sketches/template/ directory with reference template files
   - Includes multi-file example structure with helper files and library organization
   - Provides VisualGDB project files for easier development

3. **Configuration Updates**:
   - Updated examples/configs/user_settings_arduino.h to include AES CTR support
   - Added WOLFSSL_AES and WOLFSSL_AES_COUNTER macros
   - Updated version identifier to v5.7.6
   - Added ECC and RSA hardening options

4. **Arduino Integration Improvements**:
   - Added wolfssl-arduino.cpp with Arduino-specific Serial.print helper function
   - Updated wolfssl.h to properly declare the helper function
   - Updated wolfssl-arduino.sh script to include the new cpp file in Arduino library builds

5. **Documentation Updates**:
   - Updated IDE/ARDUINO/README.md with references to new examples
   - Updated IDE/ARDUINO/sketches/README.md to include new examples
   - Added troubleshooting information for common Arduino build issues

6. **Code Cleanup**:
   - Fixed file locations and directory references
   - Removed stray tabs and improved formatting
   - Fixed spelling errors in documentation and comments

Fixes zd# 19275

# Testing

How did you test?

Tested on Arduino using Teensy 4.1 

- ARM Cortex-M7 at 600 MHz
- Float point math unit, 64 & 32 bits
- 7936K Flash, 1024K RAM (512K tightly coupled), 4K EEPROM (emulated)
- QSPI memory expansion, locations for 2 extra RAM or Flash chips

See https://www.pjrc.com/store/teensy41.html

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

gojimmypi commented:
Jenkins retest this please

to retry "Unable to create live FilePath for wolf-linux-cloud-node"

gojimmypi commented:
Edit: I added a new updated file: `examples/configs/user_settings_arduino.h` that includes these macros for the AES_CTR example that would otherwise need to be added manually. 

```
#define WOLFSSL_AES
#define WOLFSSL_AES_COUNTER
``` 

This `user_settings_arduino.h` file will be the new [default user_settings.h](https://github.com/wolfSSL/Arduino-wolfSSL/blob/main/src/user_settings.h) for the _next_ release of the [Arduino-wolfSSL](https://github.com/wolfSSL/Arduino-wolfSSL) library. In the meantime, the changes can be made manually or with the [install script](Arduino-wolfSSL).

I also included some minor changes to avoid warnings in Arduino related to hardening and not using client nor server examples.

gojimmypi commented:
Jenkins retest this please.

To retry unrelated DTLS 1.3 group 6 failure 4740 "invalid stream header: 636F7272"

dgarske commented on examples/configs/user_settings_arduino.h:
```diff
@@ -25,7 +25,10 @@
+#define WOLFSSL_USER_SETTINGS_ID "Arduino user_settings.h v5.7.6"
+
+/* Disable wolfcrypt cryptographic security hardening. Comment out to enable: */
+#define WC_NO_HARDEN
```
I don't support any examples that define this... Hardening should always be enabled unless the build is public operations only.

gojimmypi commented:
Jenkins retest this please.

For "Unable to create live FilePath for wolf-linux-cloud-node-nnn; wolf-linux-cloud-node-nnn was marked offline: Connection was broken"

@dgarske I've expanded the scope of this PR to include an additional example. See ZD 19275. Updated description and PR title.

In particular note the new additional gating changes to `tls.c` that otherwise gave compiler warnings depending on client / server settings.

gojimmypi commented:
Jenkins retest this please.

multi-test-script 5141 unavailable.

gojimmypi commented:
Jenkins retest this please.

no output (or no permissions?) to test 4416. 

java.io.StreamCorruptedException: invalid stream header in 7660

dgarske reviewed (changes requested):
Is it possible to put these example into wolfssl-examples or another public repo?

gojimmypi commented:
Yes, it's certainly possible, although that would not have been my first choice. I get it that wolfSSL could contain only the core library, and arguably no examples at all.

It should be possible to modify the Arduino [publishing script](/IDE/ARDUINO/wolfssl-arduino.sh) to pull from multiple repositories (core + examples) and push everything to the [Arduino-wolfSSL](https://github.com/wolfssl/Arduino-wolfSSL) repo for publishing to Arduino (or a local directory).

Do you prefer that all or only some of the Arduino examples be moved?

dgarske commented on src/tls.c:
```diff
@@ -4777,6 +4771,7 @@ static word16 TLSX_PointFormat_Write(PointFormat* list, byte* output)
 
     return offset;
 }
+#endif /* NO_WOLFSSL_CLIENT*/
```
Please explain how to reproduce these unused function warnings. Let's get test covered on it...

gojimmypi replied:
The problem was manifested when both `NO_WOLFSSL_CLIENT` and `NO_WOLFSSL_SERVER` were defined: In my case, for the stand-alone AES-CTR Arduino example.

I removed the TLSX gating from this PR and will create a separate PR on that topic.

dgarske commented:
@gojimmypi in my opinion its better to split out library changes like TLSX gating and this Arduino feature to separate PR's. I also think you should not have the Arduino examples in the library proper. They should be in a different repo like wolfssl-examples.

gojimmypi commented:
@dgarske I've created https://github.com/wolfSSL/wolfssl-examples/pull/499 to move the Arduino examples. 

It will be a multi-step process. I'll next remove the examples from this repo in this PR, leaving only the `README.md` files that are known to be included in docs, blogs, other PRs, etc. 

Once the `wolfssl-examples` PR is merged, I will create a new PR here to modify the [install script](https://github.com/wolfSSL/wolfssl/blob/master/IDE/ARDUINO/wolfssl-arduino.sh) to pull the examples from new location when publishing to [Arduino-wolfSSL](https://github.com/wolfSSL/Arduino-wolfSSL).

Once _this_ PR is merged, I'll create a new PR in the `wolfssl-examples` to enable more testing in the new [arduino.yml](`wolfssl-examples` ) workflow file that would otherwise fail without the changes in this PR.

gojimmypi commented:
Jenkins retest this please.

For ERROR: Issue with creating launcher for agent wolf-linux-cloud-node

gojimmypi commented:
Jenkins retest this please.

for AgentOfflineException: Unable to create live FilePath for wolf-linux-cloud-node

gojimmypi commented:
Jenkins retest this please.

for: Cannot contact wolf-linux-cloud-node

dgarske commented on src/internal.c:
```diff
@@ -2589,6 +2589,28 @@ int InitSSL_Ctx(WOLFSSL_CTX* ctx, WOLFSSL_METHOD* method, void* heap)
+            /* Unless a known board is detected, assume there's no network */
+            #define WOLFSSL_NO_NETWORK
+        #else
+            ctx->CBIORecv = EmbedReceive;
```
You are duplicating all the code below for platforms that do support network IO. 

I'd prefer to see something in settings.h that detects this and sets WOLFSSL_USER_IO in the cases where there is no network.

What is `WOLFSSL_NO_NETWORK` for anyways? That doesn't exist in our code base. Please just gate on existing WOLFSSL_USER_IO.

gojimmypi replied:
> You are duplicating all the code below for platforms that do support network IO.

Yes. That was intentional. I purposely pulled out the Arduino-specific code to isolate it from defaults that might later change, even though it was currently a duplicate of the default. See update that uses default for Arduino CB.

> What is WOLFSSL_NO_NETWORK for anyways? 

That was going to be a macro for Arduino boards known to have no networking capabilities. I've removed it.

gojimmypi commented:
Hi @dgarske I've applied your suggested changes.

I've also updated the Arduino install script to now look for an optional `WOLFSSL_EXAMPLES_ROOT` environment variable that will typical be the path to the local [wolfssl-examples](https://github.com/wolfSSL/wolfssl-examples/), once https://github.com/wolfSSL/wolfssl-examples/pull/499 is merged.

There was also a duplicate `#define WOLFSSL_USER_IO` that I removed from `settings.h`.

dgarske commented:
Retest this please: Generic config PRB-single-flag.txt_105: "Found unhandled org.jenkinsci.plugins.workflow.support.steps.AgentOfflineException exception:"

gojimmypi commented on src/internal.c:
```diff
@@ -2589,6 +2589,21 @@ int InitSSL_Ctx(WOLFSSL_CTX* ctx, WOLFSSL_METHOD* method, void* heap)
+         !defined(ARDUINO_SEEED_XIAO) && !defined(ARDUINO_TEENSY41)     && \
+         !defined(WIFI_AVAILABLE)     && !defined(ETHERNET_AVAILABLE)
+        /* Unless a known board is detected, assume there's no network */
+        #ifndef NO_WOLFSSL_CLIENT
```
Perhaps I don't understand the full impact of those macros.

My _intention_ is that if all of these are true:

- Arduino build is encountered
- None of the boards known to have networking have been detected
- Macros `NO_WOLFSSL_CLIENT` and/or `NO_WOLFSSL_SERVER` are not defined

Then a compile-time warning is given that there's likely no networking and that both `NO_WOLFSSL_CLIENT` and `NO_WOLFSSL_SERVER`  are expected to be defined. 

Without those, then some of the Arduino libraires like `WiFi.h` will give many unintuitive errors at compile time.

Open to suggestions on how to better implement this.

dgarske replied:
`#if !defined(WOLFCRYPT_ONLY) && defined(NO_WOLFSSL_CLIENT) && defined(NO_WOLFSSL_SERVER)`
If its not wolfCrypt only and both no client and no server are set then you are welcome to produce a warning... This is NOT Arduino specific and could be a general warning in settings.h. Understand?

gojimmypi replied:
Ah yes, I see what you mean. I agree that's a good idea and I've added it to `settings.h`.

That is still _different_ than my intention of the _Arduino specific_ board detection of 

- "if not a known board with networking then `NO_WOLFSSL_CLIENT` and `NO_WOLFSSL_SERVER` should be defined." 

And as you pointed out, `WOLFCRYPT_ONLY` should also be defined. 

I do agree it does not belong in here in `internal.c`, and may even be beyond the scope of this PR. 

There's probably also a better Arduino way of doing this, recorded here for future reference:

```c
    #if !defined(ARDUINO_ARCH_ESP32) && !defined(ARDUINO_ARCH_ESP8266) && \
        !defined(ARDUINO_ARCH_SAMD)  && !defined(ARDUINO_ARCH_NRF52)   && \
        !defined(ARDUINO_ARCH_MBED)  && !defined(ARDUINO_ARCH_STM32)   && \
        !defined(ARDUINO_SEEED_XIAO) && !defined(ARDUINO_TEENSY41)     && \
        !defined(WIFI_AVAILABLE)     && !defined(ETHERNET_AVAILABLE)
        /* Unless a known board is detected, assume there's no network */
        #ifndef NO_WOLFSSL_CLIENT
            #warning "No known networking on this device."
            #warning "Please add NO_WOLFSSL_CLIENT to your user_settings.h"
        #endif
        #ifndef NO_WOLFSSL_SERVER
            #warning "No known networking on this device."
            #warning "Please add NO_WOLFSSL_SERVER to your user_settings.h"
        #endif
    #endif
```

gojimmypi commented on examples/configs/user_settings_arduino.h:
```diff
@@ -76,17 +84,39 @@
+    /* By default all examples are enabled; no specific optimizations */
+
+    /* Optionally warn if server or client is disabled */
+    #if 0
```
It was reference code to warn during initialization that either `NO_WOLFSSL_CLIENT` and/or `NO_WOLFSSL_SERVER` were defined, which can lead to undesired, unintuitive troubleshooting. 

Not much use if it needs to be enabled, and by the time someone knows that, they already know what to look for.

Removed.

dgarske commented:
Retest this please: multiple "Found unhandled org.jenkinsci.plugins.workflow.support.steps.AgentOfflineException exception:".

gojimmypi commented on wolfssl/wolfcrypt/settings.h:
```diff
@@ -4416,6 +4416,13 @@ extern void uITRON4_free(void *p) ;
 #endif /* HAVE_ENTROPY_MEMUSE */
 
+/* If not server and not client, WOLFCRYPT_ONLY should be defined */
+#if defined(NO_WOLFSSL_CLIENT) && defined(NO_WOLFSSL_SERVER) && \
```
ok, changed `#warning` to `#error` and squashed commits

dgarske replied:
I'm fine with this error but please also add `!defined(NO_TLS)`.

dgarske replied:
No, just one error and one section that includes `!WOLFCRYPT_ONLY && !NO_TLS`

dgarske replied:
````#if defined(NO_WOLFSSL_CLIENT) && defined(NO_WOLFSSL_SERVER) && \
    !defined(WOLFCRYPT_ONLY) && !defined(NO_TLS)
    #error "If TLS is enabled please make sure either client or server is enabled."
#endif``

dgarske commented on wolfssl/wolfcrypt/settings.h:
```diff
@@ -4416,6 +4416,13 @@ extern void uITRON4_free(void *p) ;
+/* If not server and not client, WOLFCRYPT_ONLY should be defined */
+#if defined(NO_WOLFSSL_CLIENT) && defined(NO_WOLFSSL_SERVER) && \
+   !defined(WOLFCRYPT_ONLY)
+    #error "ERROR: Both NO_WOLFSSL_CLIENT and NO_WOLFSSL_SERVER defined. "\
```
`If TLS is enabled please make sure either client or server is enabled`

gojimmypi replied:
yes, I can add that too. What's a good way to answer "_is TLS enabled_"?

`CONFIG_TLS_STACK_WOLFSSL` || `WOLFSSL_DTLS` || `HAVE_TLS_EXTENSIONS` || `WOLFSSL_TLS13` || `WOLFSSL_LEANTLS`  ?

Or is TLS _always_ enabled unless `NO_TLS` is found?

gojimmypi replied:
Hi @dgarske I've introduced new logic to check both the `WOLFCRYPT_ONLY` and `NO_TLS`  along with the `NO_WOLFSSL_CLIENT` and `NO_WOLFSSL_SERVER` with appropriate error messages.

The `NO_TLS`  was a bit more interesting for me. Upon adding to my `user_setting.h` a variety of other compiler warnings were encountered. I appeased most of them, but perhaps all of the  

`#if !defined(WOLFSSL_NO_TLS12)` 

should be updated to 

`#if !defined(NO_TLS) && !defined(WOLFSSL_NO_TLS12)`

gojimmypi commented:
> No, just one error and one section that includes !WOLFCRYPT_ONLY && !NO_TLS

Yes, I had that at one point, but when the `WOLFCRYPT_ONLY` was  missing, the error was not very intuitive.

In any case, I've applied your exact change as exactly specified.

dgarske commented on wolfssl/wolfcrypt/settings.h:
```diff
@@ -4416,6 +4416,12 @@ extern void uITRON4_free(void *p) ;
     #endif
 #endif /* HAVE_ENTROPY_MEMUSE */
 
+/* If not server and not client, WOLFCRYPT_ONLY and NO_TLS should be defined */
```
This comment makes no sense. Please remove or update it

dgarske commented on src/internal.c:
```diff
@@ -24827,6 +24832,8 @@ int SendCertificateRequest(WOLFSSL* ssl)
 
     return ret;
 }
+#endif /* NO_TLS */
```
`!NO_TLS`

gojimmypi replied:
Good catch. Corrected that one & applied some other edits, added closing gate comments, etc.

Note added file: [src/tls13.c](https://github.com/wolfSSL/wolfssl/pull/8514/files#diff-927d62a3fce45ef1b7d9ba70df049ab4dc21ea53fc2491b5c006fc96c5af680e)

gojimmypi commented:
Jenkins retest this please

for Unable to create live FilePath for wolf-linux-cloud-node-nnn;  marked offline: Connection was broken
